store/v2/root/store.go

```diff
@@ -0,0 +1,938 @@
+type viewSubstore struct {
+	dataBucket           dbm.DBReader
+	indexBucket          dbm.DBReader
+	stateCommitmentStore *smt.Store
+}
+
+// Builder type used to create a valid schema with no prefix conflicts
+type prefixRegistry struct {
+	StoreSchema
+	reserved []string
+}
+
+// Mixin types that will be composed into each distinct root store variant type
+type listenerMixin struct {
+	listeners map[types.StoreKey][]types.WriteListener
+}
+
+type traceMixin struct {
+	TraceWriter  io.Writer
+	TraceContext types.TraceContext
+}
+
+// DefaultStoreConfig returns a RootStore config with an empty schema, a single backing DB,
+// pruning with PruneDefault, no listeners and no tracer.
+func DefaultStoreConfig() StoreConfig {
+	return StoreConfig{
+		Pruning: types.PruneDefault,
+		prefixRegistry: prefixRegistry{
+			StoreSchema: StoreSchema{},
+		},
+		listenerMixin: &listenerMixin{
+			listeners: map[types.StoreKey][]types.WriteListener{},
+		},
+		traceMixin: &traceMixin{
+			TraceWriter:  nil,
+			TraceContext: nil,
+		},
+	}
+}
+
+// Returns true for valid store types for a RootStore schema
+func validSubStoreType(sst types.StoreType) bool {
+	switch sst {
+	case types.StoreTypePersistent:
+		return true
+	case types.StoreTypeMemory:
+		return true
+	case types.StoreTypeTransient:
+		return true
+	default:
+		return false
+	}
+}
+
+// Returns true iff both schema maps match exactly (including mem/tran stores)
+func (this StoreSchema) equal(that StoreSchema) bool {
+	if len(this) != len(that) {
+		return false
+	}
+	for key, val := range that {
+		myval, has := this[key]
+		if !has {
+			return false
+		}
+		if val != myval {
+			return false
+		}
+	}
+	return true
+}
+
+// Parses a schema from the DB
+func readSavedSchema(bucket dbm.DBReader) (*prefixRegistry, error) {
+	ret := prefixRegistry{StoreSchema: StoreSchema{}}
+	it, err := bucket.Iterator(nil, nil)
+	if err != nil {
+		return nil, err
+	}
+	for it.Next() {
+		value := it.Value()
+		if len(value) != 1 || !validSubStoreType(types.StoreType(value[0])) {
+			return nil, fmt.Errorf("invalid mapping for store key: %v => %v", it.Key(), value)
+		}
+		ret.StoreSchema[string(it.Key())] = types.StoreType(value[0])
+		ret.reserved = append(ret.reserved, string(it.Key())) // assume iter yields keys sorted
+	}
+	if err = it.Close(); err != nil {
+		return nil, err
+	}
+	return &ret, nil
+}
+
+// NewStore constructs a RootStore directly from a database.
+// Creates a new store if no data exists; otherwise loads existing data.
+func NewStore(db dbm.DBConnection, opts StoreConfig) (ret *Store, err error) {
+	versions, err := db.Versions()
+	if err != nil {
+		return
+	}
+	// If the DB is not empty, attempt to load existing data
+	if saved := versions.Count(); saved != 0 {
+		if opts.InitialVersion != 0 && versions.Last() < opts.InitialVersion {
+			return nil, fmt.Errorf("latest saved version is less than initial version: %v < %v",
+				versions.Last(), opts.InitialVersion)
+		}
+	}
+	err = db.Revert()
```

Comment:
why do we `Revert` here? Please add a comment.

Reply:
We revert here in order to clear any DB data that was committed but not part of a saved version. I'll add a comment and a test case.